lib/Sema/TypeCheckGeneric.cpp

```diff
@@ -1162,6 +1162,13 @@ void TypeChecker::validateGenericTypeSignature(GenericTypeDecl *typeDecl) {
 
   prepareGenericParamList(gp, typeDecl);
 
+  // For a protocol, compute the requirement signature first. It will be used
+  // by clients of the protocol.
+  if (auto proto = dyn_cast<ProtocolDecl>(typeDecl)) {
```

Comment:
That's correct; `prepareGenericParamList` sets the depth of the generic parameters, which has to happen first. Then we compute the requirement signature before building the environment.